test/Main.purs

```diff
@@ -2,10 +2,183 @@ module Test.Main where
+import PureScript.Backend.Optimizer.Convert (BackendModule)
+import PureScript.Backend.Optimizer.CoreFn (Comment(..), Module(..), ModuleName(..))
+import PureScript.Backend.Optimizer.Directives (parseDirectiveFile)
+import PureScript.Backend.Optimizer.Directives.Defaults (defaultDirectives)
+import PureScript.Backend.Optimizer.Semantics.Foreign (coreForeignSemantics)
+import Test.Utils (bufferToUTF8, coreFnModulesFromOutput, execWithStdin, mkdirp, spawnFromParent)
+
+type TestArgs =
+  { accept :: Boolean
+  , filter :: NonEmptyArray String
+  }
+
+argParser :: ArgParser TestArgs
+argParser =
+  ArgParser.fromRecord
+    { accept:
+        ArgParser.flag [ "--accept", "-a" ]
+          "Accepts snapshot output"
+          # ArgParser.boolean
+          # ArgParser.default false
+    , filter:
+        ArgParser.argument [ "--filter", "-f" ]
+          "Filter tests matching a prefix"
+          # ArgParser.unfolded1
+          # ArgParser.default (pure "Snapshot.*")
+    }
 
 main :: Effect Unit
 main = do
-  log "🍝"
-  log "You should add some tests."
+  cliArgs <- Array.drop 2 <$> Process.argv
+  case ArgParser.parseArgs "test" "" argParser cliArgs of
+    Left err ->
+      Console.error $ ArgParser.printArgError err
+    Right args ->
+      launchAff_ $ runSnapshotTests args
+
+runSnapshotTests :: TestArgs -> Aff Unit
+runSnapshotTests { accept, filter } = do
+  liftEffect $ Process.chdir $ Path.concat [ "test-snapshots" ]
+  spawnFromParent "spago" [ "build", "-u", "-g corefn" ]
+  snapshotDir <- liftEffect Process.cwd
+  snapshotPaths <- expandGlobs (Path.concat [ snapshotDir, "snapshots-input" ])
+    [ "Snapshot.*.purs" ]
+  outputRef <- liftEffect $ Ref.new Map.empty
+  let snapshotsOut = Path.concat [ snapshotDir, "snapshots-output" ]
+  let testOut = Path.concat [ snapshotDir, "test-out" ]
+  mkdirp snapshotsOut
+  mkdirp testOut
+  coreFnModulesFromOutput "output" filter >>= case _ of
+    Left errors -> do
+      for_ errors \(Tuple filePath err) -> do
+        Console.error $ filePath <> " " <> err
+      liftEffect $ Process.exit 1
+    Right coreFnModules -> do
+      let { directives } = parseDirectiveFile defaultDirectives
+      -- No runtime .ss files needed yet
+      -- copyFile (Path.concat [ "..", "..", "runtime.js" ]) (Path.concat [ testOut, "runtime.js" ])
+      coreFnModules # buildModules
+        { directives
+        , foreignSemantics: coreForeignSemantics -- no chez scheme specific foreign semantics yet
+        , onCodegenModule: \_ (Module { name: ModuleName name, path }) backend -> do
+            let
+              formatted =
+                Dodo.print plainText Dodo.twoSpaces $ S.printChezExpr $ codegenModule
+                  backend
+            let testFileDir = Path.concat [ testOut, name ]
+            let testFilePath = Path.concat [ testFileDir, "output.ss" ]
+            mkdirp testFileDir
+            FS.writeTextFile UTF8 testFilePath formatted
```

Comment:
We could, but I think that should be done in the codegen itself, right?

Reply:
Or, maybe I should say, the AST used should be modified to better indicate what's a list and what's an actual module, so that the printer can add the newline at the end of the module.

Regardless, this has been done in latest commits.

Reply:
Yeah, I'll add a `ChezFile` that's exactly that